net/nimble/transport/uart/src/ble_hci_uart.c

```diff
@@ -297,17 +321,36 @@ ble_hci_uart_tx_char(void *arg)
         /* Copy the first unsent byte from the tx buffer and remove it from the
          * source.
          */
-        os_mbuf_copydata(ble_hci_uart_state.tx_acl, 0, 1, &u8);
-        os_mbuf_adj(ble_hci_uart_state.tx_acl, 1);
-
-        /* Free the tx buffer if this is the last byte to send. */
-        if (OS_MBUF_PKTLEN(ble_hci_uart_state.tx_acl) == 0) {
-            os_mbuf_free_chain(ble_hci_uart_state.tx_acl);
+        u8 = ble_hci_uart_state.tx_pkt.dptr[0];
+        --ble_hci_uart_state.rem_tx_len;
+        if (ble_hci_uart_state.rem_tx_len == 0) {
+            os_mbuf_free_chain(ble_hci_uart_state.tx_pkt.tx_acl);
             ble_hci_uart_state.tx_type = BLE_HCI_UART_H4_NONE;
+        } else {
+            om = ble_hci_uart_state.tx_pkt.tx_acl;
+            --om->om_len;
+            if (om->om_len == 0) {
+                /* Remove and free any zero mbufs */
+                while ((om != NULL) && (om->om_len == 0)) {
+                    ble_hci_uart_state.tx_pkt.tx_acl = SLIST_NEXT(om, om_next);
+                    os_mbuf_free(om);
+                    om = ble_hci_uart_state.tx_pkt.tx_acl;
+                }
+                /* NOTE: om should never be NULL! What to do? */
```

Comment:
if om should never be NULL then maybe change while() loop above to  do{...}while(om->om_len) and have assert for this?  this will make it easier to debug issues and also check in loop is simpler